edgelet/edgelet-kube/src/module/create.rs

```diff
@@ -0,0 +1,150 @@
+use futures::prelude::*;
+use futures::{future, Future, Stream};
+use hyper::service::Service;
+use hyper::Body;
+
+use edgelet_docker::DockerConfig;
+use kube_client::{Error as KubeClientError, TokenSource};
+
+use crate::constants::EDGE_EDGE_AGENT_NAME;
+use crate::convert::{spec_to_deployment, spec_to_role_binding, spec_to_service_account};
+use crate::error::Error;
+use crate::runtime::KubeRuntimeData;
+use crate::KubeModuleRuntime;
+use edgelet_core::ModuleSpec;
+
+pub fn create_module<T, S>(
+    runtime: &KubeModuleRuntime<T, S>,
+    module: &ModuleSpec<DockerConfig>,
+) -> impl Future<Item = (), Error = Error>
+where
+    T: TokenSource + Send + 'static,
+    S: Service + Send + 'static,
+    S::ReqBody: From<Vec<u8>>,
+    S::ResBody: Stream,
+    Body: From<S::ResBody>,
+    S::Error: Into<KubeClientError>,
+    S::Future: Send,
+{
+    let runtime_for_sa = runtime.clone();
+    let module_for_sa = module.clone();
+
+    let runtime_for_deployment = runtime.clone();
+    let module_for_deployment = module.clone();
+
+    let inner = create_or_update_service_account(&runtime, &module)
+        .and_then(move |_| create_or_update_role_binding(&runtime_for_sa, &module_for_sa))
+        .and_then(move |_| {
+            create_or_update_deployment(&runtime_for_deployment, &module_for_deployment)
+        });
+
+    Box::new(inner)
+}
+
+fn create_or_update_service_account<T, S>(
+    runtime: &KubeModuleRuntime<T, S>,
+    module: &ModuleSpec<DockerConfig>,
+) -> impl Future<Item = (), Error = Error>
+where
+    T: TokenSource + Send + 'static,
+    S: Send + Service + 'static,
+    S::ReqBody: From<Vec<u8>>,
+    S::ResBody: Stream,
+    Body: From<S::ResBody>,
+    S::Error: Into<KubeClientError>,
+    S::Future: Send,
+{
+    let f = spec_to_service_account(runtime, module)
+        .map_err(Error::from)
+        .map(|(name, new_service_account)| {
+            let namespace = runtime.namespace().to_owned();
+            runtime
+                .client()
+                .lock()
+                .expect("Unexpected lock error")
+                .borrow_mut()
+                .replace_service_account(&namespace, &name, &new_service_account)
+                .map_err(Error::from)
+                .map(|_| ())
+        })
+        .into_future()
+        .flatten();
+
+    Box::new(f)
+}
+
+fn create_or_update_role_binding<T, S>(
+    runtime: &KubeModuleRuntime<T, S>,
+    module: &ModuleSpec<DockerConfig>,
+) -> impl Future<Item = (), Error = Error>
+where
+    T: TokenSource + Send + 'static,
+    S: Send + Service + 'static,
+    S::ReqBody: From<Vec<u8>>,
+    S::ResBody: Stream,
+    Body: From<S::ResBody>,
+    S::Error: Into<KubeClientError>,
+    S::Future: Send,
+{
+    let f = spec_to_role_binding(runtime, module)
+        .map_err(Error::from)
+        .map(|(name, new_role_binding)| {
+            // create new role only for edge agent
+            if new_role_binding.metadata.as_ref().unwrap().name
```

Comment:
Should we correctly handle the error here? An unwrap can cause a crash.

Reply:
Inside `spec_to_role_binding` I make sure to assign some value to metadata. But there is another issue in this line. I don't use the `name` part of tuple

Reply:
We should avoid an unwrap though.